synapse/handlers/stats.py

```diff
@@ -80,6 +80,17 @@ async def _unsafe_process(self) -> None:
         # If self.pos is None then means we haven't fetched it from DB
         if self.pos is None:
             self.pos = await self.store.get_stats_positions()
+            room_max_stream_ordering = self.store.get_room_max_stream_ordering()
+            if self.pos > room_max_stream_ordering:
+                # apparently, we've processed more events than exist in the database!
+                # this can happen if events are removed with history purge or similar.
+                logger.warning(
+                    "Event stream ordering appears to have gone backwards (%i -> %i): "
+                    "rewinding stats processor",
+                    self.pos,
+                    room_max_stream_ordering,
+                )
+                self.pos = room_max_stream_ordering
```

Comment:
I'm having trouble convincing myself this is a "safe" operation to do. I thought the max stream ordering was monotonically increasing, even if you purged events but I might be misunderstanding how it is used?

Reply:
well, that's a fair question.

The fact is that currently, on sqlite, we *do* reuse stream orderings, because we have no way of knowing which stream orderings were used before synapse was restarted. 

For things like the stats and userdirectory, that's not really a problem, because any events that get a stream ordering later than the max at startup must be new events, so I think this must be safe.

One place it *is* a problem is in client /sync - since stream_orderings are encoded in sync tokens, any events that reuse old stream_orderings probably won't get synced to the client. Ironically, we *don't* log any warnings in that case.

So, sure, a better solution here is not to reuse stream orderings at all, but that's not easy. Meanwhile I think we may as well update these bits to stop them complaining.